### PR TITLE
WICKET-6888 ToggleableDataTable

### PR DESCRIPTION
**Pull request by rkurucz9**
I would like to add a ToggleableDataTable to the codebase.  Each row of it has an expansion content.
If I have to fix my code, please share it with me.

### CODE REVIEW

File: wicket-examples/src/main/java/org/apache/wicket/examples/repeater/repeater.css [hunk 1]
```diff
@@ -72,3 +72,16 @@ col.yellow {
 col.brown {
 	background-color: #CC6633;
 }
+
+.toggleable {
```
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/repeater/repeater.css:76]: See my comment about CSS class names below.

**Review by martin-g — APPROVED** [review]
LGTM!

File: wicket-extensions/src/main/java/org/apache/wicket/extensions/markup/html/repeater/data/table/toggle/ToggleColumn.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public void populateItem(final Item<ICellPopulator<T>> cellItem, final String componentId, final IModel<T> rowModel) {
+        final Component toggleComponent = newToggleComponent(componentId, rowModel);
+        toggleComponent.add(new ClassAttributeModifier() {
+            @Override
+            protected Set<String> update(Set<String> oldClasses) {
+                final T object = rowModel.getObject();
+                if (dataTable.isToggleable(object)) {
+                    oldClasses.add("toggleable");
```
[the remark this answers is not in the document]
Comment (rkurucz9) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/markup/html/repeater/data/table/toggle/ToggleColumn.java:55]: Would I fix the other CSS classes (collapsible, expandable) in the ToggleColumn?
Would I fix it in the ToggleableDataTable class for the "expansion-row" CSS class?

[the remark this answers is not in the document]
Comment (martin-g) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/markup/html/repeater/data/table/toggle/ToggleColumn.java:55]: Yes, please!

[the remark this answers is not in the document]
Comment (rkurucz9) [reply wicket-extensions/src/main/java/org/apache/wicket/extensions/markup/html/repeater/data/table/toggle/ToggleColumn.java]: Fixed it.

File: wicket-examples/src/main/java/org/apache/wicket/examples/repeater/repeater.css (hunk 1 above)
Comment (rkurucz9) [reply wicket-examples/src/main/java/org/apache/wicket/examples/repeater/repeater.css]: I used a more specific CSS class rule for toggleable, collapsible, and expandable.

**Comment by svenmeier** [comment]
I've build a lot of these components over the years for different projects, and each had to be different according to the needs of the current project.

There are a whole lot of issues with this implementation, from less important ones like documentation and coding style to larger conceptual errors, performance considerations and code duplication.

I'd rather not pull this code chunk into Wicket as is, and restructuring it might take a long time.

**Comment by rkurucz9** [comment]
It's all right from my side. I agree with your concept and arguments.

**Comment by reiern70** [comment]
> I've build a lot of these components over the years for different projects, and each had to be different according to the needs of the current project.
> 
> There are a whole lot of issues with this implementation, from less important ones like documentation and coding style to larger conceptual errors, performance considerations and code duplication.
> 
> I'd rather not pull this code chunk into Wicket as is, and restructuring it might take a long time.

I fully agree. IMHO. this does not belong to core wicket. At least not in its current state. If cleaned, interfaces and concepts clearly defined it can be maybe added to wicket-stuff. I also had to create similar components for customer projects and I do not think it is that easy to roll a "fit all use cases component".

**Side note**: when creating these kinds of PR it is very useful to attach some images thus that reviewers can easily see how this component looks like.

Thanks for your efforts.

**Comment by svenmeier** [comment]
I've taken a second look and actually your solution is very good. You clearly have understood how to work with components :).

My major grieve is how the collection of expanded objects is held in the component tree: this defies the 'holy mantra of detachment', i.e. the selected entities will get serialized along with the page. You could take a look at AbstractTree#newModel() for inspiration how this can be avoided.

**Comment by rkurucz9** [comment]
> I've taken a second look and actually your solution is very good. You clearly have understood how to work with components :).
> 
> My major grieve is how the collection of expanded objects is held in the component tree: this defies the 'holy mantra of detachment', i.e. the selected entities will get serialized along with the page. You could take a look at AbstractTree#newModel() for inspiration how this can be avoided.

Thank you for your praising and suggestion. :) I try to fix my solution according to the AbstractTree#newModel() aspect.

**Comment by rkurucz9** [comment]
> I've taken a second look and actually your solution is very good. You clearly have understood how to work with components :).
> 
> My major grieve is how the collection of expanded objects is held in the component tree: this defies the 'holy mantra of detachment', i.e. the selected entities will get serialized along with the page. You could take a look at AbstractTree#newModel() for inspiration how this can be avoided.

I hope I could fix it properly.

**Pull request closed without merging**